java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/buildsteps/RuntimeImageBuildStep.java

```diff
@@ -17,57 +17,93 @@
 package com.google.cloud.runtimes.builder.buildsteps;
 
 import static com.google.cloud.runtimes.builder.Constants.DOCKERFILE_BUILD_STAGE;
+import static com.google.cloud.runtimes.builder.config.domain.Artifact.ArtifactType.APP_ENGINE_EXPLODED_WAR;
+import static com.google.cloud.runtimes.builder.config.domain.Artifact.ArtifactType.EXPLODED_WAR;
+import static com.google.cloud.runtimes.builder.config.domain.Artifact.ArtifactType.JAR;
+import static com.google.cloud.runtimes.builder.config.domain.Artifact.ArtifactType.WAR;
 
 import com.google.cloud.runtimes.builder.buildsteps.base.BuildStep;
 import com.google.cloud.runtimes.builder.buildsteps.base.BuildStepException;
+import com.google.cloud.runtimes.builder.config.domain.Artifact;
 import com.google.cloud.runtimes.builder.config.domain.BuildContext;
 import com.google.cloud.runtimes.builder.config.domain.JdkServerLookup;
 import com.google.cloud.runtimes.builder.config.domain.RuntimeConfig;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 public abstract class RuntimeImageBuildStep implements BuildStep {
 
+  private final Logger logger = LoggerFactory.getLogger(RuntimeImageBuildStep.class);
+
   private final JdkServerLookup jdkServerLookup;
+  private final String compatImageName;
 
-  protected RuntimeImageBuildStep(JdkServerLookup jdkServerLookup) {
+  protected RuntimeImageBuildStep(JdkServerLookup jdkServerLookup, String compatImageName) {
     this.jdkServerLookup = jdkServerLookup;
+    this.compatImageName = compatImageName;
   }
 
   @Override
   public void run(BuildContext buildContext) throws BuildStepException {
-    buildContext.getDockerfile().appendLine("FROM " + getBaseRuntimeImage(buildContext));
+    Artifact artifact = getArtifact(buildContext);
+    logger.debug("Found Java artifact {}", artifact);
 
-    String copyStep = "COPY ";
+    buildContext.getDockerfile().appendLine("FROM " + getBaseRuntimeImage(buildContext, artifact));
+    String copyStep = "COPY";
     if (buildContext.isSourceBuild()) {
-      copyStep += "--from=" + DOCKERFILE_BUILD_STAGE + " ";
+      copyStep += " --from=" + DOCKERFILE_BUILD_STAGE;
     }
-    buildContext.getDockerfile().appendLine(copyStep + getArtifact(buildContext)
-        + " $APP_DESTINATION");
+
+    String relativeArtifactPath = "./" + buildContext.getWorkspaceDir()
+        .relativize(artifact.getPath()).toString();
+
+    // compat runtime requires a special app destination
+    String artifactDestination = artifact.getType() == APP_ENGINE_EXPLODED_WAR
```

Comment:
I spent some time looking into this, and it's actually a bit tricky to get exploded wars working in the jetty runtime. Since this would be a new feature beyond what's available in XRT, do you have any objections to creating a separate issue for this to be tackled later? Until then I can remove the non-compat EXPLODED_WAR ArtifactType.